File: 2024/Day05/OrderingUpdates.cpp
#include <bits/stdc++.h>
using namespace std;

bool isCorrectOrder( const vector<int>& update, const map<int, vector<int>>& rules ) {
    for( int i = 0; i < update.size(); i++ ) {
        for( int j = 0; j < i; j++ ) {
            for( int after : rules.at( update[i] ) ) {
                if( after == update[j] ) {
                    return false;
                }
            }
        }
    }
    return true;
}

map<int, vector<int>> readRules() {
    map<int, vector<int>> rules;
    FILE* RuleInput = fopen( "RuleInput.txt", "r" );

    char linebuf[BUFSIZ];
    while( fgets( linebuf, BUFSIZ, RuleInput ) ) {
        stringstream ss( linebuf );
        int num1 = 0, num2 = 0; char sep = '\0';
        ss >> num1 >> sep >> num2;
        rules[num1].push_back( num2 );
    }
    return move( rules );

}

vector<vector<int>> readUpdates() {
    vector<vector<int>> updates;
    char linebuf[BUFSIZ];
    FILE* UpdatesInput = fopen( "UpdatesInput.txt", "r" );
    while( fgets( linebuf, BUFSIZ, UpdatesInput ) ) {
        stringstream ss( linebuf );
        int tmp = 0;
        vector<int> update;
        while( ss >> tmp ) {
            update.push_back( tmp );
            char comma = '\0';
            if( ( ss >> comma ) && comma == ',' ) {
                continue;
            }
        }
        if( update.size() > 0 )
            updates.push_back( update );
    }
    return move( updates );
}


void Solution1() {
    map<int, vector<int>> rules = readRules();
    vector<vector<int>> updates = readUpdates();
    int sum = 0;
    for( const auto& update : updates ) {
        if( isCorrectOrder( update, rules ) ) {
            int middle = update.size() / 2;
            sum += update[middle];
        }
    }
    cout << "The sum of the middle page numbers of the correctly reordered updates is: " << sum << endl;
}

void reOrder( vector<int>& update, const map<int, vector<int>>& rules ) {
    for( int i = 0; i < update.size(); i++ ) {
        for( int j = i; j < update.size(); j++ ) {
            bool ahead = true;
            for( int k = i; k < update.size(); k++ ) {
                if( k == j ) { continue; }
                for( auto after : rules.at( update[k] ) ) {
                    if( after == update[j] ) { ahead = false; break; }
                }
                if( !ahead ) break;
            }
            if( ahead ) {
                swap( update[i], update[j] );
                break;
            }
        }
    }
}

int Solution2() {
    map<int, vector<int>> rules = readRules();
    vector<vector<int>> updates = readUpdates();
    int sum = 0;
    for( vector<int> update : updates ) {
        if( isCorrectOrder( update, rules ) ) {
            int middle = update.size() / 2;
        }
        else {
            reOrder( update, rules );
            int middle = update.size() / 2;
            sum += update[middle];
        }
    }
    cout << "The sum of the middle page numbers of the correctly reordered updates is: " << sum << endl;
    return 0;
}

int main() {
    Solution1();
    Solution2();
    return 0;
}

File: 2024Java/Day05/OrderingUpdates.java
package Day05;

import java.io.*;
import java.util.*;
import java.util.stream.Collectors;

public class OrderingUpdates {

    HashMap<Integer, Set<Integer>> readRules() throws IOException {
        HashMap<Integer, Set<Integer>> rules = new HashMap<>();
        BufferedReader RuleInput = new BufferedReader(new FileReader("RuleInput.txt"));
        String linebuf;
        while ((linebuf = RuleInput.readLine()) != null) {
            int num1 = Integer.parseInt(linebuf.split("\\|")[0]),
                    num2 = Integer.parseInt(linebuf.split("\\|")[1]);
            rules.putIfAbsent(num1, new HashSet<>());
            rules.get(num1).add(num2);
        }
        RuleInput.close();
        return rules;
    }

    List<List<Integer>> readUpdates() throws IOException {
        List<List<Integer>> updates = new ArrayList<>();
        String linebuf;
        BufferedReader UpdatesInput = new BufferedReader(new FileReader("UpdatesInput.txt"));
        while ((linebuf = UpdatesInput.readLine()) != null) {
            if (!linebuf.isEmpty()) {
                List<Integer> update = Arrays.asList(linebuf.trim().split(",")).stream()
                        .map(l -> Integer.parseInt(l)).collect(Collectors.toList());

                updates.add(update);
            }
        }
        UpdatesInput.close();
        return updates;
    }

    boolean isCorrectOrder(List<Integer> update, HashMap<Integer, Set<Integer>> rules) {
        for (int i = 0; i < update.size(); i++) {
            for (int j = 0; j < i; j++) {
                if (rules.getOrDefault(update.get(i), null) != null
                        && rules.get(update.get(i)).contains(update.get(j))) {
                    return false;
                }
            }
        }
        return true;
    }

    void Solution1() throws IOException {
        HashMap<Integer, Set<Integer>> rules = readRules();
        List<List<Integer>> updates = readUpdates();
        int sum = 0;
        for (var update : updates) {
            if (isCorrectOrder(update, rules)) {
                int middle = update.size() / 2;
                sum += update.get(middle);
            }
        }
        System.out.println("Solution 1: " + sum);
    }

    void reOrder(List<Integer> update, HashMap<Integer, Set<Integer>> rules) {
        for (int i = 0; i < update.size(); i++) {
            for (int j = i; j < update.size(); j++) {
                boolean ahead = true;
                for (int k = i; k < update.size(); k++) {
                    if (k == j) {
                        continue;
                    }
                    if (rules.getOrDefault(update.get(k), null) != null
                            && rules.get(update.get(k)).contains(update.get(j))) {
                        ahead = false;
                        break;
                    }
                }
                if (ahead) {
                    var t = update.get(i);
                    update.set(i, update.get(j));
                    update.set(j, t);
                    break;
                }
            }
        }
    }

    int Solution2() throws IOException {
        HashMap<Integer, Set<Integer>> rules = readRules();
        List<List<Integer>> updates = readUpdates();
        int sum = 0;
        for (List<Integer> update : updates) {
            if (!isCorrectOrder(update, rules)) {
                reOrder(update, rules);
                int middle = update.size() / 2;
                sum += update.get(middle);
            }
        }
        System.out.println("Solution 2: " + sum);
        return 0;
    }

    public static void main(String[] args) throws IOException {
        OrderingUpdates Day05 = new OrderingUpdates();
        Day05.Solution1();
        Day05.Solution2();
    }
}
